### ISSUE TITLE
Some slt queries cause integer overflow

### ISSUE THREAD

**Issue by jamii** [MaterializeInc/materialize#1651]
This probably happened when we changed the default int type to Int32. We're only able to see these when running in debug mode, so CI didn't catch it.

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.34s
     Running `target/debug/sqllogictest test/sqlite/test/random/expr/slt_good_0.test`
==> test/sqlite/test/random/expr/slt_good_0.test
SELECT ALL - CASE COUNT ( DISTINCT + 2 ) WHEN + 13 THEN - 79 * - 91 * - 73 * 67 * 94 + - + CAST ( NULL AS INTEGER ) ELSE NULL END * - - 69 + + 39 + - COUNT ( DISTINCT 8 )
thread '<unnamed>' panicked at 'attempt to multiply with overflow', src/expr/scalar/func.rs:666:17
stack backtrace:
   0:     0x55d81a4cf0b4 - backtrace::backtrace::libunwind::trace::h65597d255cb1398b
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1:     0x55d81a4cf0b4 - backtrace::backtrace::trace_unsynchronized::hd4f479d7150ec4a0
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2:     0x55d81a4cf0b4 - std::sys_common::backtrace::_print_fmt::h015072984a2b172c
                               at src/libstd/sys_common/backtrace.rs:77
   3:     0x55d81a4cf0b4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6df05d3335f32194
                               at src/libstd/sys_common/backtrace.rs:61
   4:     0x55d81a4f3adc - core::fmt::write::h1f444f4312eb6c27
                               at src/libcore/fmt/mod.rs:1028
   5:     0x55d81a4c9687 - std::io::Write::write_fmt::h8d147888220078ef
                               at src/libstd/io/mod.rs:1412
   6:     0x55d81a4d188e - std::sys_common::backtrace::_print::h8a6df0fa81d6af62
                               at src/libstd/sys_common/backtrace.rs:65
   7:     0x55d81a4d188e - std::sys_common::backtrace::print::h6f05b4733407e509
                               at src/libstd/sys_common/backtrace.rs:50
   8:     0x55d81a4d188e - std::panicking::default_hook::{{closure}}::h0d0a23bd02315dd8
                               at src/libstd/panicking.rs:188
   9:     0x55d81a4d1581 - std::panicking::default_hook::h8d15a9aecb4efac6
                               at src/libstd/panicking.rs:205
  10:     0x55d819f6cd38 - <alloc::boxed::Box<F> as core::ops::function::Fn<A>>::call::h2f1d0f65db915e60
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/liballoc/boxed.rs:956
  11:     0x55d819f6cb1c - ore::panic::set_abort_on_panic::{{closure}}::h0267835260f2d54c
                               at src/ore/panic.rs:31
  12:     0x55d81a4d2060 - std::panicking::rust_panic_with_hook::hbe174577402a475d
                               at src/libstd/panicking.rs:468
  13:     0x55d81a4d1b2e - std::panicking::continue_panic_fmt::h4d855dad868accf3
                               at src/libstd/panicking.rs:373
  14:     0x55d81a4d1a16 - rust_begin_unwind
                               at src/libstd/panicking.rs:302
  15:     0x55d81a4ef85e - core::panicking::panic_fmt::hdeb7979ab6591473
                               at src/libcore/panicking.rs:139
  16:     0x55d81a4ef7aa - core::panicking::panic::hb5daa85c7c72fc62
                               at src/libcore/panicking.rs:70
  17:     0x55d819c2353e - expr::scalar::func::mul_int32::hfc6dc67aaa4d52ca
                               at src/expr/scalar/func.rs:666
  18:     0x55d819c28b56 - expr::scalar::func::BinaryFunc::eval::hcfb5b7fdd6a67b74
                               at src/expr/scalar/func.rs:1515
  19:     0x55d819b836f0 - expr::scalar::ScalarExpr::eval::h4ca832377834174d
                               at src/expr/scalar/mod.rs:398
  20:     0x55d819d24046 - expr::scalar::ScalarExpr::reduce::{{closure}}::h6266cec7d7136fbd
                               at src/expr/scalar/mod.rs:244
  21:     0x55d819d242d0 - expr::scalar::ScalarExpr::reduce::{{closure}}::h03d27efe1111b9e7
                               at src/expr/scalar/mod.rs:258
  22:     0x55d819d237b4 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:159
  23:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  24:     0x55d819d23510 - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:135
  25:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  26:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  27:     0x55d819d23551 - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:138
  28:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  29:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  30:     0x55d819d236cc - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:148
  31:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  32:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  33:     0x55d819d23551 - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:138
  34:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  35:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  36:     0x55d819d23510 - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:135
  37:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  38:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  39:     0x55d819d23551 - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:138
  40:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  41:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  42:     0x55d819d23510 - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:135
  43:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  44:     0x55d819d23c84 - expr::scalar::ScalarExpr::visit_mut::{{closure}}::h36ff68e10137ccf9
                               at src/expr/scalar/mod.rs:158
  45:     0x55d819d23551 - expr::scalar::ScalarExpr::visit1_mut::hffab1a0e95e10a46
                               at src/expr/scalar/mod.rs:138
  46:     0x55d819d23798 - expr::scalar::ScalarExpr::visit_mut::h4004cb7f050e880f
                               at src/expr/scalar/mod.rs:158
  47:     0x55d819b82ad1 - expr::scalar::ScalarExpr::reduce::h3bc067e0d23682e6
                               at src/expr/scalar/mod.rs:246
  48:     0x55d819db15ad - expr::transform::reduction::FoldConstants::action::h6b09654118e9afae
                               at src/expr/transform/reduction.rs:131
  49:     0x55d819dd81c0 - expr::transform::reduction::FoldConstants::transform::{{closure}}::h6072fdb62b6bc60b
                               at src/expr/transform/reduction.rs:34
  50:     0x55d819c151e4 - expr::relation::RelationExpr::visit_mut::h86baa67dc37e2344
                               at src/expr/relation/mod.rs:714
  51:     0x55d819c15584 - expr::relation::RelationExpr::visit_mut::{{closure}}::hd193fb049c0100fb
                               at src/expr/relation/mod.rs:713
  52:     0x55d819c0ed6f - expr::relation::RelationExpr::visit1_mut::h3ce9dec11bfdee38
                               at src/expr/relation/mod.rs:674
  53:     0x55d819c151c8 - expr::relation::RelationExpr::visit_mut::h86baa67dc37e2344
                               at src/expr/relation/mod.rs:713
  54:     0x55d819c15584 - expr::relation::RelationExpr::visit_mut::{{closure}}::hd193fb049c0100fb
                               at src/expr/relation/mod.rs:713
  55:     0x55d819c0ed3c - expr::relation::RelationExpr::visit1_mut::h3ce9dec11bfdee38
                               at src/expr/relation/mod.rs:671
  56:     0x55d819c151c8 - expr::relation::RelationExpr::visit_mut::h86baa67dc37e2344
                               at src/expr/relation/mod.rs:713
  57:     0x55d819db08c4 - expr::transform::reduction::FoldConstants::transform::h5f98cfeda7579042
                               at src/expr/transform/reduction.rs:33
  58:     0x55d819db087d - <expr::transform::reduction::FoldConstants as expr::transform::Transform>::transform::h7784a7d9541453e2
                               at src/expr/transform/reduction.rs:27
  59:     0x55d819cff7fd - <expr::transform::Optimizer as expr::transform::Transform>::transform::hda598a3bc9dfc255
                               at src/expr/transform/mod.rs:93
  60:     0x55d819d01477 - expr::transform::Optimizer::optimize::h84c825aeaa6539a0
                               at src/expr/transform/mod.rs:185
  61:     0x55d817b61794 - coord::coord::Coordinator<C>::optimize::h67e3109ae8797400
                               at /home/jamie/materialize/src/coord/coord.rs:896
  62:     0x55d817b50b49 - coord::coord::Coordinator<C>::sequence_plan::hdc79e6769c751d65
                               at /home/jamie/materialize/src/coord/coord.rs:594
  63:     0x55d817b60f70 - coord::coord::Coordinator<C>::serve::{{closure}}::{{closure}}::h580b692aea4dc908
                               at /home/jamie/materialize/src/coord/coord.rs:314
  64:     0x55d817bfc890 - core::result::Result<T,E>::and_then::h4913745358c8e7cc
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/result.rs:691
  65:     0x55d817b5f472 - coord::coord::Coordinator<C>::serve::{{closure}}::h8eb11b205174349e
                               at /home/jamie/materialize/src/coord/coord.rs:313
  66:     0x55d818802bd9 - tokio::runtime::context::enter::h212ef260a28d42fd
                               at /home/jamie/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.9/src/runtime/context.rs:72
  67:     0x55d8179cf485 - tokio::runtime::handle::Handle::enter::hd863a9619c68cc57
                               at /home/jamie/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.9/src/runtime/handle.rs:33
  68:     0x55d817b5e8d3 - coord::coord::Coordinator<C>::serve::h8e310e5fcf8e958f
                               at /home/jamie/materialize/src/coord/coord.rs:285
  69:     0x55d8176f8b58 - sqllogictest::runner::State::start::{{closure}}::h71322064422696a2
                               at src/sqllogictest/runner.rs:391
  70:     0x55d818671535 - std::sys_common::backtrace::__rust_begin_short_backtrace::h4593907358c292b7
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/sys_common/backtrace.rs:129
  71:     0x55d817ee4c14 - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::hdd54fa35c0b25bcb
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/thread/mod.rs:469
  72:     0x55d8187480c4 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h26346aff0ad2ab3c
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/panic.rs:317
  73:     0x55d81890a341 - std::panicking::try::do_call::h9cf26bcbcddbff4e
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/panicking.rs:287
  74:     0x55d81a4d6eaa - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:78
  75:     0x55d8189040a8 - std::panicking::try::h25db22b1a8833caf
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/panicking.rs:265
  76:     0x55d818749166 - std::panic::catch_unwind::hb583bf2b22d01b4f
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/panic.rs:396
  77:     0x55d817ee464d - std::thread::Builder::spawn_unchecked::{{closure}}::h41d591423706247f
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/thread/mod.rs:468
  78:     0x55d817218a54 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h304adf91934c0b4a
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/ops/function.rs:227
  79:     0x55d81a4c23ef - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h3534c64212330b0c
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/liballoc/boxed.rs:942
  80:     0x55d81a4d5fd0 - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h338c10574a337ece
                               at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/liballoc/boxed.rs:942
  81:     0x55d81a4d5fd0 - std::sys_common::thread::start_thread::h761ac6d57710d65d
                               at src/libstd/sys_common/thread.rs:13
  82:     0x55d81a4d5fd0 - std::sys::unix::thread::Thread::new::thread_start::h61c012ef60f933c0
                               at src/libstd/sys/unix/thread.rs:79
  83:     0x7f0b27512ef7 - start_thread
  84:     0x7f0b2723222f - __GI___clone
  85:                0x0 - <unknown>
Aborted (core dumped)
```

**Comment by ruchirK** [comment]
I ran into this today on a fuzzer generated query

```
SELECT (CAST(-6734743351254754377 AS bigint) * CAST(-997833593175988522 AS bigint)) AS ident_11;
```

crashes the dev build, and in release it returns

```
materialize=> SELECT (CAST(-6734743351254754377 AS bigint) * CAST(-997833593175988522 AS bigint)) AS ident_11;
       ident_11
----------------------
 -2604240165483522822
(1 row)
```

psql handles this with

```

Test=# SELECT (CAST(-6734743351254754377 AS bigint) * CAST(-997833593175988522 AS bigint)) AS ident_11;
ERROR:  bigint out of range
Test=#
```

**Comment by cjubb39** [comment]
This specific issue appears to be fixed by https://github.com/MaterializeInc/materialize/commit/bb52a327129204652a2bc514a383092a9da317e5.  I ran ruchir's sample query in the materialize console and properly errored.

However, we have a similar issue for floats here: https://github.com/MaterializeInc/materialize/issues/8527.

When exploring around here, I ran into an issue where we're unable to use numbers that are outside of the `numeric` range but within the range of a `float`: https://github.com/MaterializeInc/materialize/issues/8529
